File: app/src/main/java/com/boardgamegeek/service/SyncUploadTask.java
package com.boardgamegeek.service;

import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.v4.app.NotificationCompat;
import android.support.v4.app.NotificationCompat.Action;
import android.support.v4.app.NotificationCompat.Builder;
import android.text.TextUtils;

import com.boardgamegeek.R;
import com.boardgamegeek.io.BggService;
import com.boardgamegeek.util.LargeIconLoader;
import com.boardgamegeek.util.LargeIconLoader.Callback;
import com.boardgamegeek.util.NotificationUtils;
import com.boardgamegeek.util.PresentationUtils;

import java.util.ArrayList;
import java.util.List;

import hugo.weaving.DebugLog;
import timber.log.Timber;

public abstract class SyncUploadTask extends SyncTask {
	private final List<CharSequence> notificationMessages = new ArrayList<>();

	@DebugLog
	public SyncUploadTask(Context context, BggService service) {
		super(context, service);
	}

	@StringRes
	protected abstract int getNotificationTitleResId();

	protected abstract Intent getNotificationIntent();

	protected abstract String getNotificationMessageTag();

	protected abstract String getNotificationErrorTag();

	@DebugLog
	protected void notifyUser(final CharSequence title, final CharSequence message, final int id, String imageUrl, final String thumbnailUrl) {
		LargeIconLoader loader = new LargeIconLoader(context, imageUrl, thumbnailUrl, new Callback() {
			@Override
			public void onSuccessfulIconLoad(Bitmap bitmap) {
				buildAndNotify(title, message, id, bitmap);
			}

			@Override
			public void onFailedIconLoad() {
				buildAndNotify(title, message, id, null);
			}
		});
		loader.executeInBackground();

		notificationMessages.add(PresentationUtils.getText(context, R.string.msg_play_upload, title, message));
		showNotificationSummary();
	}

	private void buildAndNotify(CharSequence title, CharSequence message, int id, Bitmap largeIcon) {
		Builder builder = createNotificationBuilder()
			.setCategory(NotificationCompat.CATEGORY_SERVICE)
			.setContentTitle(title)
			.setContentText(message)
			.setLargeIcon(largeIcon)
			.setOnlyAlertOnce(true)
			.setGroup(getNotificationMessageTag());
		NotificationCompat.BigTextStyle detail = new NotificationCompat.BigTextStyle(builder);
		detail.bigText(message);
		Action action = createMessageAction();
		if (action != null) {
			builder.addAction(action);
		}
		if (largeIcon != null) {
			builder.extend(new NotificationCompat.WearableExtender().setBackground(largeIcon));
		}
		NotificationUtils.notify(context, getNotificationMessageTag(), id, builder);
	}

	@DebugLog
	private void showNotificationSummary() {
		Builder builder = createNotificationBuilder()
			.setGroup(getNotificationMessageTag())
			.setGroupSummary(true);
		final int messageCount = notificationMessages.size();
		if (messageCount == 1) {
			builder.setContentText(notificationMessages.get(0));
		} else {
			NotificationCompat.InboxStyle detail = new NotificationCompat.InboxStyle(builder);
			for (int i = messageCount - 1; i >= 0; i--) {
				detail.addLine(notificationMessages.get(i));
			}
		}
		NotificationUtils.notify(context, getNotificationMessageTag(), 0, builder);
	}

	@Nullable
	@DebugLog
	protected Action createMessageAction() {
		return null;
	}

	@DebugLog
	protected void notifyUploadError(CharSequence errorMessage) {
		if (TextUtils.isEmpty(errorMessage)) return;
		Timber.e(errorMessage.toString());
		NotificationCompat.Builder builder = createNotificationBuilder()
			.setContentText(errorMessage)
			.setCategory(NotificationCompat.CATEGORY_ERROR);
		NotificationCompat.BigTextStyle detail = new NotificationCompat.BigTextStyle(builder);
		detail.bigText(errorMessage);
		NotificationUtils.notify(context, getNotificationErrorTag(), 0, builder);
	}

	@DebugLog
	private NotificationCompat.Builder createNotificationBuilder() {
		return NotificationUtils.createNotificationBuilder(context,
			context.getString(getNotificationTitleResId()),
			getNotificationIntent());
	}
}

File: app/src/main/java/com/boardgamegeek/service/SyncUploadTask.kt
package com.boardgamegeek.service

import android.content.Intent
import android.content.SyncResult
import android.graphics.Bitmap
import androidx.annotation.PluralsRes
import androidx.annotation.StringRes
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationCompat.Action
import com.boardgamegeek.BggApplication
import com.boardgamegeek.R
import com.boardgamegeek.extensions.*
import com.boardgamegeek.util.LargeIconLoader
import com.boardgamegeek.util.LargeIconLoader.Callback
import timber.log.Timber

abstract class SyncUploadTask(application: BggApplication, syncResult: SyncResult) : SyncTask(application, syncResult) {
    private val notificationMessages = ArrayList<CharSequence>()

    @get:StringRes
    protected abstract val notificationTitleResId: Int

    protected abstract val notificationSummaryIntent: Intent

    protected open val notificationIntent: Intent?
        get() = notificationSummaryIntent

    protected abstract val notificationMessageTag: String

    protected abstract val notificationErrorTag: String

    @get:PluralsRes
    protected abstract val summarySuffixResId: Int

    protected fun notifyUser(title: CharSequence, message: CharSequence, id: Int, vararg imageUrl: String) {
        if (prefs[KEY_SYNC_UPLOADS, true] != true) return

        notificationMessages.add(context.getText(R.string.msg_play_upload, title, message))

        val loader = LargeIconLoader(context, *imageUrl, callback = object : Callback {
            override fun onSuccessfulIconLoad(bitmap: Bitmap) {
                buildAndNotify(title, message, id, bitmap)
            }

            override fun onFailedIconLoad() {
                buildAndNotify(title, message, id)
            }
        })
        loader.executeInBackground()
    }

    private fun buildAndNotify(title: CharSequence, message: CharSequence, id: Int, largeIcon: Bitmap? = null) {
        val builder = context.createNotificationBuilder(
            notificationTitleResId,
            NotificationChannels.SYNC_UPLOAD,
            notificationIntent
        )
            .setCategory(NotificationCompat.CATEGORY_SERVICE)
            .setContentTitle(title)
            .setContentText(message)
            .setLargeIcon(largeIcon)
            .setOnlyAlertOnce(true)
            .setGroup(notificationMessageTag)
            .setGroupAlertBehavior(NotificationCompat.GROUP_ALERT_SUMMARY)
            .setStyle(NotificationCompat.BigTextStyle().bigText(message))
        val action = createMessageAction()
        if (action != null) {
            builder.addAction(action)
        }
        context.notify(builder, notificationMessageTag, id)
        showNotificationSummary()
    }

    private fun showNotificationSummary() {
        val builder = context.createNotificationBuilder(
            notificationTitleResId,
            NotificationChannels.SYNC_UPLOAD,
            notificationSummaryIntent
        )
            .setGroup(notificationMessageTag)
            .setGroupAlertBehavior(NotificationCompat.GROUP_ALERT_SUMMARY)
            .setGroupSummary(true)
        when (val messageCount = notificationMessages.size) {
            0 -> return
            1 -> builder.setContentText(notificationMessages[0])
            else -> {
                val message = context.resources.getQuantityString(summarySuffixResId, notificationMessages.size, notificationMessages.size)
                builder.setContentText(message)
                val detail = NotificationCompat.InboxStyle(builder).setSummaryText(message)
                for (i in messageCount - 1 downTo 0) {
                    detail.addLine(notificationMessages[i])
                }
            }
        }
        context.notify(builder, notificationMessageTag)
    }

    protected open fun createMessageAction(): Action? {
        return null
    }

    protected fun notifyUploadError(errorMessage: CharSequence) {
        if (errorMessage.isBlank()) return
        Timber.e(errorMessage.toString())
        val builder = context.createNotificationBuilder(notificationTitleResId, NotificationChannels.ERROR, notificationSummaryIntent)
            .setContentText(errorMessage)
            .setCategory(NotificationCompat.CATEGORY_ERROR)
        val detail = NotificationCompat.BigTextStyle(builder)
        detail.bigText(errorMessage)
        context.notify(builder, notificationErrorTag)
    }
}
